pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py

```diff
@@ -0,0 +1,189 @@
+# -*- coding: utf-8 -*-
+import enum
+import json
+import os
+import sys
+from typing import List, Literal, Tuple, TypedDict, Union
+
+import pytest
+
+
+# Inherit from str so it's JSON serializable.
+class TestNodeTypeEnum(str, enum.Enum):
+    class_ = "class"
+    file = "file"
+    folder = "folder"
+    test = "test"
+
+
+class TestData(TypedDict):
+    name: str
+    path: str
+    type_: TestNodeTypeEnum
+    id_: str
+
+
+class TestItem(TestData):
+    lineno: str
+    runID: str
+
+
+class TestNode(TestData):
+    children: "List[TestNode | TestItem]"
+
+
+# Add the path to pythonFiles to sys.path to find testing_tools.socket_manager.
+PYTHON_FILES = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+sys.path.insert(0, PYTHON_FILES)
+
+# Add the lib path to sys.path to find the typing_extensions module.
+sys.path.insert(0, os.path.join(PYTHON_FILES, "lib", "python"))
+from testing_tools import socket_manager
+from typing_extensions import NotRequired
+
+DEFAULT_PORT = "45454"
+
+
+def pytest_collection_finish(session):
+    node, error = build_test_tree(session)
+    cwd = os.getcwd()
+    # TODO: add error checking.
+    sendPost(cwd, node)
+
+
+def build_test_tree(session) -> Tuple[Union[TestNode, None], List[str]]:
+    errors: List[str] = []
+    session_node = create_session_node(session)
+    # a dictionary of all direct children of the session.
+    session_children_dict: dict[str, TestNode] = dict()
+    # a dictionary of all files in the session.
+    file_nodes_dict: dict[pytest.Module, TestNode] = dict()
+    # a dictionary of all classes in the session.
+    class_nodes_dict: dict[str, TestNode] = dict()
+    # iterate through all the test items in the session.
+    for test_case in session.items:
+        test_node = create_test_node(test_case)
+        # Check parent node type, either Module or UnitTest class.
+        if type(test_case.parent) == pytest.Module:
+            file_nodes_dict.setdefault(
+                test_case.parent, create_file_node(test_case.parent)
+            )["children"].append(test_node)
+        else:
+            test_class_node = class_nodes_dict.setdefault(
+                test_case.parent.name,
+                create_class_node(test_case.parent),
+            )
+            test_class_node["children"].append(test_node)
```

Comment:
How is this code different than the lines above? Is it just the dict you are working with? If so then I would assign to a new variable that is assigned based on the conditional guard and then uses the same code.